File: openframeworks/lib/mesh/Follower.cpp
//
//  Follower.cpp
//  Persistence
//
//  Created by Mike Chambers on 4/19/16.
//
//

#include "Follower.h"

Follower::Follower() : Mover() {
    
}

Follower::Follower(Mover * _target) : Mover() {
    target = _target;
}

Follower::Follower(const Follower &source) {
    target = new Mover();
    *target = *source.target; // copy the value
    
    attractionCoefficient = source.attractionCoefficient;
    index = source.index;
}

void Follower::setTarget(Mover * _target) {
    target = _target;
}

void Follower::setIndex(int _index) {
    index = _index;
}

void Follower::update() {
    update(target->location);
}

void Follower::update(ofVec3f t) {
    
    //do we need this limit?
    velocity.limit(limit);
    ofVec3f dir = t - location;
    dir.normalize();
    dir *= attractionCoefficient;
    
    //this used to be =, which would reset acceleration, ignoring
    //any changes through apply force
    acceleration += dir;
    velocity += acceleration;
    location += velocity;
    
    updateAngle();
    acceleration *= 0;
}

File: paperjs/_scripts/mesh/Follower.js
/*jslint vars: true, nomen: true, plusplus: true, continue:true, forin:true */
/*global atob, btoa, ArrayBuffer, Uint8Array, Blob */

class Follower extends Mover {

	//var target = null;
	//var attractionCoefficient = 0.1;

	constructor (target) {
		super();
		this.target = target;
		this.attractionCoefficient = 0.1;
	}

	set target(value) {
		this._target = value;
	}

	get target() {
		return this._target;
	}

	update(tPoint) {
    	if(!tPoint) {
    		tPoint = this.target.location;
    	}

    	//todo
    	this.velocity = this.velocity.normalize(this.limit);

    	var dir = tPoint.subtract(this.location);

    	dir = dir.normalize();

    	//+=
    	dir = dir.multiply(this.attractionCoefficient);

	    this.acceleration = this.acceleration.add(dir);
	    this.velocity = this.velocity.add(this.acceleration);
	    this.location = this.location.add(this.velocity);
	    
	    this.updateAngle();
	    this.acceleration.set(0,0);
    }
}
